src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -60,85 +47,62 @@ internal bool TryGetBuffer(in SequencePosition start, in SequencePosition end, o
 
                 data = new ReadOnlyMemory<T>(Unsafe.As<T[]>(startObject));
             }
-            else if (type == SequenceType.MemoryManager)
-            {
-                Debug.Assert(startObject is MemoryManager<T>);
-
-                data = (Unsafe.As<MemoryManager<T>>(startObject)).Memory;
-            }
             else if (typeof(T) == typeof(char) && type == SequenceType.String)
             {
                 Debug.Assert(startObject is string);
 
                 data = (ReadOnlyMemory<T>)(object)(Unsafe.As<string>(startObject)).AsMemory();
             }
-            else
+            else // if (type == SequenceType.MemoryManager)
             {
-                data = default;
-                return false;
+                Debug.Assert(startObject is MemoryManager<T>);
+
+                data = (Unsafe.As<MemoryManager<T>>(startObject)).Memory;
             }
 
-            data = data.Slice(startIndex, length);
+            data = data.Slice(startIndex, endIndex - startIndex);
             return true;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal ReadOnlyMemory<T> GetFirstBuffer(in SequencePosition start, in SequencePosition end)
+        private ReadOnlyMemory<T> GetFirstBuffer(in SequencePosition start, in SequencePosition end)
         {
-            SequenceType type;
-            int startIndex = 0;
-            int endIndex = 0;
+            GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
             object startObject = start.GetObject();
-            if (startObject != null)
```

Comment:
Why don't we need to check `startObject != null` anymore?

Reply:
For the last segment TryGet method return true and return Position with GetObject() equal to null. It is used as flag that all segments was processed. So when TryGet called with Position with GetObject() == nul means that all segments was processed and we return false. It is only one condition to return false and stop loop processing.